Add coriolis matrix computation

haudren opened the pull request:
This PR adds the work I did while I was still at the lab: it allows to compute the full Coriolis matrix (i.e. `C(q, qd)` instead of `C(q, qd)qd`).

I know it is not perfect, but I would like to put it out there !

There might also be some overlap with #46 as I had to implement a "block expand" functionality to obtain decent performance.

Things to fix:
- [x] Indentation should use tabs
- [x] Tests should be a real test, not just a print
- [x] Windows build is probably broken

Maybes (what's your opinion @gergondet ):
- [x] Make a Coriolis benchmark
- [x] Split the Jacobian expand logic from the Coriolis side
- [x] Add python bindings

haudren committed:
Add coriolis matrix computation

haudren committed:
Remove useless utilities

haudren committed:
Make real tests out of pre-existant tests

haudren committed:
[LINT] Add braces around initializer lists

haudren committed:
Fix indentation

haudren commented:
I used the following `clang-format` config to fix my indentation, might be useful to put it somewhere in the project.

```yaml
BasedOnStyle: LLVM
IndentWidth: 8
UseTab: Always
BreakBeforeBraces: Allman
AllowShortIfStatementsOnASingleLine: false
IndentCaseLabels: false
```

haudren committed:
Remove useless prototypes

haudren committed:
[WIP] Fix windows build

haudren committed:
[WIP] Windows add array include

haudren committed:
[WIP] Windows : set seed without time

haudren committed:
[WIP] Windows build

haudren committed:
Add a bunch of comments

haudren committed:
Remove unused variable

haudren committed:
Test every style of expansion

haudren committed:
Regular expand now uses expandAdd

haudren commented:
@gergondet : Any remarks ?

gergondet commented on src/RBDyn/Coriolis.h:
```diff
@@ -0,0 +1,84 @@
+#ifndef CORIOLIS_H_KED6CAHK
+#define CORIOLIS_H_KED6CAHK
```
Please use `#pragma once`

haudren replied:
Done

gergondet commented on src/RBDyn/Coriolis.h:
```diff
@@ -0,0 +1,84 @@
+#ifndef CORIOLIS_H_KED6CAHK
```
Missing copyright header

haudren replied:
Done

gergondet commented on src/RBDyn/Coriolis.h:
```diff
@@ -0,0 +1,84 @@
+
+#include <RBDyn/MultiBodyConfig.h>
+#include <RBDyn/CoM.h>
+#include <rbdyn/config.hh>
```
Not needed since it is included by others

haudren replied:
Done

gergondet commented on src/RBDyn/Coriolis.h:
```diff
@@ -0,0 +1,84 @@
+namespace rbd
+{
+
+using Block = std::array<int, 3>;
```
`Eigen::DenseIndex` or `Eigen::Index` (>= 3.3)

haudren replied:
Done

gergondet commented on src/RBDyn/Coriolis.h:
```diff
@@ -0,0 +1,84 @@
+using Block = std::array<int, 3>;
+using Blocks = std::vector<Block>;
+
+/** Expand a symmetric product of a jacobian by its transpose onto every DoF.
```
I'm not sure I understand what you mean about the symmetric product here, `jacMat` can be anything.

haudren replied:
Well `jacMat` can be anything, but it *must* be a `dof x dof` matrix. Then, it will be expanded by blocks of `dof_j1 x dof_j2` which means that a `jacMat` not representing a `J^T J`  (symmetric product) is meaningless.

gergondet commented on src/RBDyn/Coriolis.h:
```diff
@@ -0,0 +1,84 @@
+ * @param jacMat The matrix containing the product J^T*J to be expanded
+ * @param res The accumulator matrix
+ **/
+RBDYN_DLLAPI void compactExpandAdd(const Blocks& compactPath,
```
Rename `expandAdd` and let C++ overloading do its magic

haudren replied:
Done

gergondet commented on src/RBDyn/Coriolis.h:
```diff
@@ -0,0 +1,84 @@
+ */
+class RBDYN_DLLAPI Coriolis
+{
+	public:
```
`public`/`private` alignment should be the same as the class declaration

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+
+BOOST_AUTO_TEST_CASE(CoriolisTest)
+{
+	std::srand(133757348);
```
`std::srand(std::time(0));` looks [ok on VS](https://godbolt.org/g/P6gqeT)

You should also `#include <cstdlib>` and `#include <ctime>`

haudren replied:
I just wanted the test to be random, but reproducible, hence the fixed seed. If you prefer a truly random seed, I'll change it back.

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+	rbd::MultiBodyConfig mbtc;
+	rbd::MultiBodyGraph mbtg;
+
+	std::tie(mbt, mbtc, mbtg) = makeTree30Dof(false);
```
Why do you bother with having the same robot twice? This does not look necessary and it makes the test code a bit tedious to read

haudren replied:
Indeed. This is a leftover from when I was trying various settings (with / without inertia transformed to origin, with / without floating base...).

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+				q[2] = qd.y();
+				q[3] = qd.z();
+			}
+		}
```
This could be a small function. Also, you can assume the free-flyer is `mbc.q[0]` and avoid the loop/check here

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+		rbd::forwardKinematics(mbt, mbtc);
+		rbd::forwardVelocity(mbt, mbtc);
+
+		rbd::ForwardDynamics fd(mbt);
```
You can keep this object out of the loop

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+		mbtc.zero(mbt);
+
+		Eigen::VectorXd q = Eigen::VectorXd::Random(mb.nrParams());
+		mbc.q = rbd::sVectorToParam(mb, q);
```
No need for the safe version here

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+		Eigen::VectorXd gravity = fd.C();
+
+		Eigen::VectorXd qd = Eigen::VectorXd::Random(mb.nrDof());
+		mbc.alpha = rbd::sVectorToDof(mb, qd);
```
No need for the safe version here

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+		rbd::forwardVelocity(mb, mbc);
+		rbd::forwardVelocity(mbt, mbtc);
+
+		rbd::Coriolis coriolis(mb);
```
You can keep this object out of the loop

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+
+		double dt = 1e-8;
+
+		rbd::sEulerIntegration(mbt, mbtc, dt);
```
No need for the safe version here

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+		fd.computeH(mbt, mbtc);
+		Eigen::MatrixXd m1 = fd.H();
+
+		double dt = 1e-8;
```
This is a really small timestep, why?

haudren replied:
Because we will compute finite differences on H, not perform integration for simulation. Hence, the timestep must be as small as possible.

gergondet commented on tests/ExpandTest.cpp:
```diff
@@ -0,0 +1,76 @@
+
+BOOST_AUTO_TEST_CASE(ExpandJacobianTest)
+{
+	std::srand(133757348);
```
Same comment as in the Coriolis test

haudren replied:
Done

gergondet commented on tests/CoriolisTest.cpp:
```diff
@@ -0,0 +1,104 @@
+#include <RBDyn/FK.h>
+#include <RBDyn/FV.h>
+
+#include <iostream>
```
Not needed

haudren replied:
Done

gergondet commented on tests/ExpandTest.cpp:
```diff
@@ -0,0 +1,76 @@
+#include <RBDyn/FV.h>
+#include <RBDyn/EulerIntegration.h>
+
+#include <iostream>
```
Not needed

haudren replied:
Done

gergondet commented on tests/ExpandTest.cpp:
```diff
@@ -0,0 +1,76 @@
+		mbc.zero(mb);
+
+		Eigen::VectorXd q = Eigen::VectorXd::Random(mb.nrParams());
+		mbc.q = rbd::sVectorToParam(mb, q);
```
No need for the safe version here

haudren replied:
Done

gergondet commented on tests/ExpandTest.cpp:
```diff
@@ -0,0 +1,76 @@
+				q[2] = qd.y();
+				q[3] = qd.z();
+			}
+		}
```
Same remark as Coriolis

haudren replied:
Done

gergondet commented on tests/ExpandTest.cpp:
```diff
@@ -0,0 +1,76 @@
+#include <RBDyn/FD.h>
+#include <RBDyn/FK.h>
+#include <RBDyn/FV.h>
+#include <RBDyn/EulerIntegration.h>
```
You only need/use FK/FV here

haudren replied:
Done

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+			const rbd::MultiBody& mb,
+			const Eigen::MatrixXd& jacMat,
+			Eigen::MatrixXd& res)
+{
```
Needs an assert on the size of `res`

haudren replied:
Done

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+	int start_block = mb.jointPosInDof(jac.jointsPath()[0]);
+	int len_block = mb.joint(jac.jointsPath()[0]).dof();
+
+	int start;
```
Declare inside the loop since it is not meant to be used outside

haudren replied:
Done

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+		if(i == jac.jointsPath()[0])
+		{
+			continue;
+		}
```
Replace with the more old-school but more explicit in this case:
```cpp
const auto & jPath = jac.jointsPath();
for(size_t j = 1; j < jPath().size(); ++j)
{
    int i = jPath[j];
```

haudren replied:
Fixed it, thanks

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+
+Eigen::MatrixXd Coriolis::coriolis(const rbd::MultiBody& mb, const rbd::MultiBodyConfig& mbc)
+{
+	Eigen::MatrixXd coriolis = Eigen::MatrixXd::Zero(mb.nrDof(), mb.nrDof());
```
We probably want to keep this allocated in the `Coriolis` object and change this function declaration to return a const reference to it.

haudren replied:
Done

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+	{
+		jacMats_[i] = jacs_[i].jacobian(mb, mbc);
+
+		jacDotMats_[i] = jacs_[i].jacobianDot(mb, mbc);
```
I don't think you need to have `jacMats_` and `jacDotMats_`, `Jacobian::jacobian` and `Jacobian::jacobianDot` do the work on the matrix they store and returns you a reference to that matrix so you could have:

```cpp
const auto & jac = jacs_[i].jacobian(mb, mbc);
const auto & jacDot = jacs_[i].jacobianDot(mb, mbc);
```

haudren replied:
Done

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+
+		double mass = mb.body(i).inertia().mass();
+		inertia = mb.body(i).inertia().inertia()
+			- sva::vector3ToCrossMatrix<double>(mass*jacs_[i].point())*sva::vector3ToCrossMatrix(jacs_[i].point()).transpose();
```
For efficiency sake you might want a specific function to return `sva::vector3ToCrossMatrix<double>(point)*sva::vector3ToCrossMatrix<double>(point).transpose()` directly

haudren replied:
Do you really think we would gain much by splitting this into a function?

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+
+		Eigen::MatrixXd res = mass*jvi.transpose()*jDvi
+				+ jwi.transpose()*((rot*ir)*jDwi
+				+  (rDot*ir)*jwi);
```
Please don't cut the line at this point, it makes it look like this does not belong in the parentheses expression

haudren replied:
Done

gergondet commented on src/Coriolis.cpp:
```diff
@@ -0,0 +1,147 @@
+				+ jwi.transpose()*((rot*ir)*jDwi
+				+  (rDot*ir)*jwi);
+
+		compactExpandAdd(compactPaths_[i], res, coriolis);
```
If `compactExpandAdd` was accepting `const Eigen::Ref<const Eigen::MatrixXd> &` instead of `const Eigen::MatrixXd &` we could allocate `res` once (at the biggest jacobian size) and pass a `res` block here. This should be benched though...

haudren replied:
See my comment.

gergondet replied:
Ok, I've done a similar patch in 6db09a6fc54bba54c2ceba028efb7ff46cffa48 plus a small improvement in expand functions (using `noalias()` on `res` which seems reasonable)

gergondet commented:
Sorry for the long delay, I was on vacation but I am back to work now!

> Make a Coriolis benchmark

Sure.

>  Split the Jacobian expand logic from the Coriolis side

Sure. That could be use as in #46 

Few things that I didn't add to the review:
- This could be part of the `Jacobian` interface
- We could store the compact path in the `Jacobian`
- I would use a light structure rather than a raw array for `Block`, this make `compactExpandAdd` hard to decipher without looking at how the compact path is computed

>  Add python bindings

Will do when the C++ version is finalized.

haudren committed:
Various fixes

haudren committed:
More fixes

* Return a const ref to result instead of allocating
* Use const ref to jacobians instead of copy
* Reduce number of rounds in test
* Formatting fixes

haudren commented:
@gergondet : I think I fixed a lot of the mentioned points, but I'm out of time for today, I'll come back to it tomorrow to fix the rest. Feel free to modify the branch as you see fit of course.

haudren commented:
Note that the build is failing because it can't find boost on Windows apparently? Weird.

haudren committed:
Use a light structure for Block

haudren commented:
I tried implementing your suggestion of using only one `res` and passing blocks of it:

Version            | Debug (100 rounds) | Release (1000 rounds)
------------------|--------------------------|-----------------------------
One allocation |  7.75s  |  0.13s
Multiple allocs  | 7.94s  | 0.15s

~~So it is *slower* to have a single allocation in Release apparently, the compiler must be doing some magic behind the scenes.~~

It is marginally faster to have a single allocation in Release.

here is the diff:
```diff
diff --git a/src/Coriolis.cpp b/src/Coriolis.cpp
index 8c2243f..1bdbcc8 100644
--- a/src/Coriolis.cpp
+++ b/src/Coriolis.cpp
@@ -62,8 +62,8 @@ Blocks compactPath(const rbd::Jacobian& jac,
 }
 
 void expandAdd(const Blocks& compactPath,
-			const Eigen::MatrixXd& jacMat,
-			Eigen::MatrixXd& res)
+		const Eigen::Ref<const Eigen::MatrixXd>& jacMat,
+		Eigen::MatrixXd& res)
 {
 	for(const auto& b1 : compactPath)
 	{
@@ -105,6 +105,14 @@ const Eigen::MatrixXd& Coriolis::coriolis(const rbd::MultiBody& mb, const rbd::M
 
         coriolis_.setZero();
 
+        int res_size = 0;
+        for(const auto& jac : jacs_)
+        {
+          res_size = std::max(res_size, jac.dof());
+        }
+
+        Eigen::MatrixXd res = Eigen::MatrixXd::Zero(res_size, res_size);
+
 	for(int i = 0; i < mb.nrBodies(); ++i)
 	{
 		const auto & jac = jacs_[i].jacobian(mb, mbc);
@@ -129,10 +137,10 @@ const Eigen::MatrixXd& Coriolis::coriolis(const rbd::MultiBody& mb, const rbd::M
 		 *        + J_{w_i}^T R_i I_i R_i^T \dot{J}_{w_i}
 		 *        + J_{w_i}^T \dot{R}_i I_i R_i^T J_{w_i} */
 
-		Eigen::MatrixXd res = mass*jvi.transpose()*jDvi
+		res.topLeftCorner(jacs_[i].dof(), jacs_[i].dof()) = mass*jvi.transpose()*jDvi
 				+ jwi.transpose()*((rot*ir)*jDwi +  (rDot*ir)*jwi);
 
-		expandAdd(compactPaths_[i], res, coriolis_);
+		expandAdd(compactPaths_[i], res.topLeftCorner(jacs_[i].dof(), jacs_[i].dof()), coriolis_);
 	}
 
 	return coriolis_;
diff --git a/src/RBDyn/Coriolis.h b/src/RBDyn/Coriolis.h
index d9512c4..904cd5e 100644
--- a/src/RBDyn/Coriolis.h
+++ b/src/RBDyn/Coriolis.h
@@ -73,7 +73,7 @@ RBDYN_DLLAPI Blocks compactPath(const rbd::Jacobian& jac,
  * @param res The accumulator matrix
  **/
 RBDYN_DLLAPI void expandAdd(const Blocks& compactPath,
-			const Eigen::MatrixXd& jacMat,
+			const Eigen::Ref<const Eigen::MatrixXd>& jacMat,
 			Eigen::MatrixXd& res);
 
 /**
diff --git a/tests/CoriolisTest.cpp b/tests/CoriolisTest.cpp
index 636d605..a1a1cc5 100644
--- a/tests/CoriolisTest.cpp
+++ b/tests/CoriolisTest.cpp
@@ -38,7 +38,7 @@ BOOST_AUTO_TEST_CASE(CoriolisTest)
 	rbd::ForwardDynamics fd(mb);
 	rbd::Coriolis coriolis(mb);
 
-	const static int ROUNDS = 100;
+	const static int ROUNDS = 1000;
 
 	for (int i = 0; i < ROUNDS; ++i)
 	{
```

haudren commented:
My bad, I forgot to change the number of rounds in the Release/Multiple Allocs case. Updated the table.

gergondet committed:
Add missing copyright notice

gergondet committed:
Correct rbd::Block doxygen

gergondet committed:
Minor performance improvment in Coriolis computation

- Keep the intermediary result matrix stored in the class
- Use noalias() in expand functions

gergondet committed:
Use google benchmark and add Coriolis benchmark

gergondet committed:
Add Coriolis bindings

gergondet committed:
Move expand and block functionalities to Jacobian

gergondet committed:
Fix a warning in IS

gergondet committed:
Implement addFullJacobian and use it in Momentum

gergondet committed:
Add a block-aware addFullJacobian and use it in Momentum

gergondet commented:
I've adressed the 3 remaining todos and used the `Block` feature in Momentum, this should now close #46 and #45 

Actually, this opens the possibility for a very small optimization in the computation of the momentum dot matrix.

I will try to figure out why the AppVeyor build is failing tomorrow...

haudren commented:
@gergondet : Nice work, thanks ! Let me know if you need anything else :+1:

haudren committed:
Add momentum matrix benchmark

haudren commented:
I also made a quick bench for momentum:

Revision | Momentum matrix | Momentum matrix dot |
------------|------------------------|------------------------------|
a1e6f0ad (master) | 5.6e-5 | 1.2e-4 |
56b5449b (PR) | 2.7e-5 | 5.5e-5| 

So we are on average about twice as fast, not bad :smile: 

@gergondet : Should I commit this benchmark in `JacobianBench` ?

gergondet committed:
[ci] Update AppVeyor script

gergondet commented:
Ok, CI should be ok now after updating the Boost version.

Will merge when all CI lands.

haudren commented:
Thanks :fireworks:
